core-bundle/src/Cache/EntityTagger.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Cache;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Doctrine\Persistence\Mapping\MappingException;
+
+class EntityTagger
+{
+    /**
+     * @var array<string, ClassMetadata>
+     */
+    private array $classMetadata = [];
+    private EntityManagerInterface $entityManager;
+
+    public function __construct(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * @param iterable|string|object|null $target
+     *
+     * @return array<int, string>
+     */
+    public function getTags($target): array
+    {
+        if (empty($target)) {
+            return [];
+        }
+
+        if (\is_string($target)) {
+            if (class_exists($target, false)) {
+                try {
+                    return [$this->getTagForEntityClass($target)];
+                } catch (\InvalidArgumentException $e) {
+                    // ignore
+                }
+            }
+
+            return [$target];
+        }
+
+        if (is_iterable($target)) {
```

Comment:
Could this lead to problems if an entity class implements `Traversable`?

Reply:
Hmm, maybe. But it's very unlikely to implement this on an entity. We could alternatively check for doctrine's `Collection` interface. Wdyt?

 The only important thing is to *not* use arrays, because then iterating and getting the id would not be lazy anymore. :slightly_smiling_face:

Reply:
> We could alternatively check for doctrine's `Collection` interface. Wdyt?

👍

Reply:
done so in  c30a621b3a8463b22f2fd5e3ddb6b2a4d0431668